### ISSUE TITLE
It would be nice to have support of pickle protocol in write_graphite plugin

### ISSUE THREAD

**Issue by antonsoroko** [collectd/collectd#2229]
Hi!
It would be nice to have support of [pickle](http://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-pickle-protocol) protocol in [write_graphite](https://collectd.org/wiki/index.php/Plugin:Write_Graphite) plugin.

This feature can help reduce load on graphite server.

But I understand that this protocol is python-oriented, and maybe it is impossible to use it in c code. (I found some c libraries for pickle protocol, but they are old).

**Comment by octo** [comment]
Agreed, would be a nice feature to have. When I looked last, the pickle protocol was not documented well enough to implement, though.

**Comment by Ferroin** [comment]
This:
http://spootnik.org/entries/2014/04/05/diving-into-the-python-pickle-formatt/
may be of some interest, as it covers implementing an interface to Graphite via the binary protocol in Clojure.  It looks like it covers everything that would be needed to write up a serializer in C.  It's about the only place outside of the Python source code that I've seen any actual documentation for the pickle data serialization format.

**Comment by antonsoroko** [comment]
Also it is possible to call Python functions from C:
http://stackoverflow.com/a/1296612
https://docs.python.org/2/extending/extending.html#calling-python-functions-from-c

But I'm not sure how convenient it is.

Also there is ancient library - https://github.com/anmcn/chutney
But it's probably useless nowadays.

**Comment by octo** [comment]
Thanks @Ferroin, that looks useful :-)

@antonsoroko the _Python plugin_ can execute arbitrary Python scripts. I don't want to open this can of worms in any other plugin.

**Comment by rubenk** [comment]
The python documentation at https://docs.python.org/3/library/pickle.html has this to say about pickle.

> Warning The pickle module is not secure against erroneous or maliciously constructed data. Never unpickle data received from an untrusted or unauthenticated source.

While strictly speaking this is a problem for the receiver, not the sender, I'm afraid that once we support sending pickled data from collectd to Graphite, more people are going to turn on pickle support in Graphite.

I think time is better spend adding a binary marshalling format to Graphite, like protobuf or messagepack.

**Comment by antonsoroko** [comment]
@rubenk 
Hi.
By the way, new version of "vanilla" carbon has protobuf support.
https://github.com/graphite-project/carbon/blob/1.0.0/conf/carbon.conf.example#L111
https://github.com/graphite-project/carbon/pull/624

**Comment by towolf** [comment]
We are using `go-carbon` now fed by collectd `write_graphite`.

This drop-in re-implementation also now supports protobuf.

**Comment by octo** [comment]
FTR, I'd be happy to accept a patch adding Protobuf formatting to the *write_graphite plugin*.
